### PR TITLE
[PyTorch][caffe2] Add CAFFE2_{DECLARE,DEFINE}_KNOWN_TYPE

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83874
* __->__ #83707

It looks like we aren't getting inlining for the defined `_typeMetaData` functions from CAFFE_KNOWN_TYPE and there's some cost associated with that. I added new macros that fix this problem; I will migrate to them in a follow-up after I get buy-in from reviewers.

Differential Revision: [D36883685](https://our.internmc.facebook.com/intern/diff/D36883685/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36883685/)!

### CODE REVIEW

**Comment by swolchok** [comment]
btw @ezyang if there is a better reviewer than you at this point for "core C++ primitives in c10/aten" I would love a reminder as to who that should be. Guessing maybe @bdhirsh ?

**Review by ezyang — APPROVED** [review]

File: c10/util/typeid.h [hunk 1]
```diff
@@ -508,12 +510,40 @@ class C10_API TypeMeta final {
...
+  // NOTE [ TypeIdentifier::Get nvcc/clang discrepancy]
+  // nvcc and clang do not produce identical results for
+  // TypeIdentifier::Get, because TypeIdentifier::Get relies on
+  // __PRETTY_FUNCTION__ and they don't agree on the canonical names
+  // of types (e.g., nvcc normalizes to `short unsigned int`, but clang
+  // calls it `unsigned short`). Hide the implementation of this function
+  // from nvcc so that we always use clang (or whatever host C++ compiler)
+  // for TypeIdentifier::Get.
+  template <class T>
+    static uint16_t addTypeMetaData();
```
Comment (swolchok) [inline c10/util/typeid.h:532]: probably need to add EXPORT_IF_NOT_GCC here to fix tests

**Comment by swolchok** [comment]
forcing merge to bypass SIGKILL test failure, @ezyang confirmed I can ignore it

**Comment by swolchok** [comment]
this is a relatively minor performance fix that is mostly of interest to deployments at scale that care about small percentages of CPU usage

**Pull request closed without merging**